core/tasks.py

```diff
@@ -386,18 +388,53 @@ def execute(self, taskID, *args, **kwargs):
 
 ## Decorators ##
 
-def noRetry(f):
-    """Prevents a deferred Function from being called a second time"""
+def retry_n_times(retries: int, email_recipients: None | str | list[str] = None):
+    """Wrapper for deferred tasks to limit the amount of retries"""
 
-    @wraps(f)
-    def wrappedFunc(*args, **kwargs):
-        try:
-            f(*args, **kwargs)
-        except Exception as e:
-            logging.exception(e)
-            raise PermanentTaskFailure()
+    def outer_wrapper(func):
+        @wraps(func)
+        def inner_wrapper(*args, **kwargs):
+            retry_count = int(currentRequest.get().request.headers.get("X-Appengine-Taskretrycount", -1))
+            try:
+                func(*args, **kwargs)
+            except Exception as exc:
+                logging.exception(f"Task {func.__qualname__} failed.")
+                logging.info("This was retry #%d. %d retries remaining. (total = %d)",
+                             retry_count, retries - retry_count, retries)
+                if retry_count < retries:
+                    # Raise the exception to mark this task as failed, so the Task Queue can retry it.
+                    raise exc
+                else:
+                    if email_recipients:
+                        try:
+                            from viur.core import email
+                            email.sendEMail(
+                                dests=email_recipients,
+                                # language=Jinja2
+                                stringTemplate="""Subject: Task {{func_name}} failed {{retries}} times\n
+                                This was the last attempt.<br>\n
+                                <pre>{{func_module|escape}}.{{func_name|escape}}({{args|map("escape")|join(", ")}}, {{kwargs.items()|map("join", "=")|join(", ")|escape}})</pre>
+                                <pre>{{traceback|escape}}</pre>""",
```

Comment:
Maybe we shoukld move this in a HTML template in viur-base the you can modify it easier

Reply:
You can now select your own template per parameter.
The default text should be present in the core, otherwise it would not work after an update.